frontend/lib/resolution/scope-queries.cpp

```diff
@@ -1576,6 +1591,123 @@ const owned<ResolvedVisibilityScope>& resolveVisibilityStmtsQuery(
+                               UniqueString formalName,
+                               std::vector<BorrowedIdsWithName>& matches) {
+  // Check that there is a match that isn't a method/field
+  // to skip the warning for collisions with secondary methods.
+  bool onlyMethodsFields = true;
+  for (auto b : matches) {
+    if (!b.containsOnlyMethodsOrFields()) {
+      onlyMethodsFields = false;
+      break;
+    }
+  }
+  if (onlyMethodsFields) {
+    return;
+  }
+
+  // find the Formal*
+  const Formal* formal = nullptr;
+  std::vector<BorrowedIdsWithName> ids;
+  functionScope->lookupInScope(formalName, ids,
+                               /* ignore private */ false,
+                               /* only methods/fields */ false);
+  for (auto b : ids) {
+    for (auto id : b) {
+      auto formalAst = parsing::idToAst(context, id);
+      if (formalAst != nullptr) {
+        formal = formalAst->toFormal();
+        break;
+      }
+    }
+  }
+
+  // find the VisibilityClause
+  const VisibilityClause* clause = nullptr;
+  if (is != nullptr) {
+    auto clauseAst = parsing::idToAst(context, is->visibilityClauseId());
+    if (clauseAst != nullptr) {
+      clause = clauseAst->toVisibilityClause();
+    }
+  }
+
+  // find if the parent is a use or import
+  VisibilityStmtKind kind = VIS_USE;
+  for (ID cur = parsing::idToParentId(context, clause->id());
+       !cur.isEmpty();
+       cur = parsing::idToParentId(context, cur)) {
+    auto tag = parsing::idToTag(context, cur);
+    if (asttags::isUse(tag)) {
+      kind = VIS_USE;
+      break;
+    } else if (asttags::isImport(tag)) {
+      kind = VIS_IMPORT;
+      break;
+    }
+  }
+
+  if (formal && clause) {
+    CHPL_REPORT(context, HiddenFormal, formal, clause, kind);
+  }
+}
+
+static const bool& warnHiddenFormalsQuery(Context* context,
+                                          const ResolvedVisibilityScope* rs,
+                                          const Scope* functionScope) {
+  QUERY_BEGIN(warnHiddenFormalsQuery, context, rs, functionScope);
+
+  bool result = false;
+
+  // warn if a function formal name conflicts with something
+  // brought in by use/import.
+
+  std::set<UniqueString> formalNames = functionScope->gatherNames();
+
+  NamedScopeSet checkedScopes;
+  std::vector<BorrowedIdsWithName> matches;
+  const VisibilitySymbols* foundVs = nullptr;
+
+  for (auto name : formalNames) {
+    bool onlyInnermost = true;
+    bool skipPrivateVisibilities = false;
+    bool got = false;
+
+    // TODO: an alternative to calling into details of the
+    // scope resolver here would be to provide a way for doLookupInScope
+    // to return the visibility clauses used in finding a symbol.
+    got = doLookupInImportsAndUses(context, rs->scope(), rs, rs, name,
+                                   onlyInnermost, skipPrivateVisibilities,
+                                   VisibilitySymbols::REGULAR_SCOPE,
+                                   checkedScopes, matches,
```

Comment:
Should checkedScopes be cleared between invocations, since it's used for different scope levels? Here and elsewhere?

Reply:
No, because we don't want to add the same symbols if the are (say) multiple `use` statements that `public use` the same module.